src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorHtmlPublishDiagnosticsInterceptor.cs

```diff
@@ -0,0 +1,195 @@
+                razorDocumentSnapshot.Version,
+                htmlDocumentSnapshot,
+                cancellationToken).ConfigureAwait(false);
+            if (!synchronized)
+            {
+                // Could not synchronize, we'll clear out the diagnostics as we don't
+                // know whether or not they're still valid in the current state of the
+                // document.
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var unmappedDiagnostics = diagnosticParams.Diagnostics;
+            var filteredDiagnostics = unmappedDiagnostics.Where(d => !CanDiagnosticBeFiltered(d));
+            if (!filteredDiagnostics.Any())
+            {
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var rangesToMap = filteredDiagnostics.Select(r => r.Range).ToArray();
+            var mappingResult = await _documentMappingProvider.MapToDocumentRangesAsync(
+                RazorLanguageKind.Html,
+                razorDocumentUri,
+                rangesToMap,
+                LanguageServerMappingBehavior.Inclusive,
+                cancellationToken).ConfigureAwait(false);
+
+            if (mappingResult == null || mappingResult.HostDocumentVersion != razorDocumentSnapshot.Version)
+            {
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var mappedDiagnostics = new List<Diagnostic>(filteredDiagnostics.Count());
+
+            for (var i = 0; i < filteredDiagnostics.Count(); i++)
+            {
+                var diagnostic = filteredDiagnostics.ElementAt(i);
+                var range = mappingResult.Ranges[i];
+
+                if (range.IsUndefined())
+                {
+                    // Couldn't remap the range correctly.
+                    // If this isn't an `Error` Severity Diagnostic we can discard it.
+                    if (diagnostic.Severity != DiagnosticSeverity.Error)
+                    {
+                        continue;
+                    }
+
+                    // For `Error` Severity diagnostics we still show the diagnostics to
+                    // the user, however we set the range to an undefined range to ensure
+                    // clicking on the diagnostic doesn't cause errors.
+                }
+
+                diagnostic.Range = range;
+                mappedDiagnostics.Add(diagnostic);
+            }
+
+            diagnosticParams.Diagnostics = mappedDiagnostics.ToArray();
+
+            return CreateResponse(diagnosticParams);
+
+
+            static InterceptionResult CreateDefaultResponse(JToken token) =>
+                new InterceptionResult(token, changedDocumentUri: false);
+
+            static InterceptionResult CreateResponse(
+                VSPublishDiagnosticParams diagnosticParams,
+                bool clearDiagnostics = false,
```

Comment:
Instead of baking the "clear" aspect into this method I'd propose another method (doesn't matter about the duplication) for clarity. Kind of like how you have "default repsonse" above.

Reply:
That's a good point, thanks, updated.